File: Java/src/main/java/QuestionBank/Solution001_100/Solution16.java
package QuestionBank.Solution001_100;

import java.util.Arrays;

/**
 * @author windows
 */
public class Solution16 {
    public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);
        int res = nums[0] + nums[1] + nums[2];
        int min = Integer.MAX_VALUE;
        for (int i = 0; i < nums.length - 2; i++) {
            int left = i + 1;
            int right = nums.length - 1;
            while (left < right) {
                if (min > Math.abs(target - nums[i] - nums[left] - nums[right])){
                    min = Math.abs(target - nums[i] - nums[left] - nums[right]);
                    res = nums[i] + nums[left] + nums[right];
                }
                if (nums[left] + nums[right] < target - nums[i]){
                    left++;
                }else if (nums[left] + nums[right] > target - nums[i]){
                    right--;
                }else{
                    return res;
                }
            }
        }
        return res;
    }


    public static void main(String[] args) {
        Solution16 solution16 = new Solution16();
        System.out.println(solution16.threeSumClosest(new int[]{1, 1, 1, 0}, -100));
    }
}

File: C++/src/Solution1_100/Solution16.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution16.h"
#include <functional>
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <string>
using namespace std;
#include "algorithm"

int Solution::threeSumClosest(vector<int> &nums, int target) {
    sort(nums.begin(), nums.end());
    int n = nums.size();
    int min_val = 1e7;
    int res;
    for (int i = 0; i < n; ++i) {
        int j = i + 1, k = n - 1;
        while (j < k){
            int num = nums[i] + nums[j] + nums[k];
            if (abs(target - num) < min_val){
                min_val = abs(target - num);
                res = num;
            }
            if (num == target){
                return res;
            } else if (num < target){
                j++;
            }else {
                k--;
            }
        }
    }
    return res;
}
